### DISCUSSION TITLE
Show keyboard shortcuts in menus and tooltips

### DISCUSSION THREAD

**Discussion by monsdar** [OpenRCT2/OpenRCT2#15762] [Feature requests]
In order to learn the keyboard shortcuts while playing it would be of great help to show the keyboard shortcuts in menus and tooltips.

**Comment by ocalhoun6** [comment]
Would this take into account any custom keyboard shortcuts? If I reassign the shortcut for something, will the tooltip change to show that?

(Perfectly fine if it doesn't -- I realize that's an uncommon case, and it would probably make the idea 10x harder to code.)

**Comment by Broxzier** [comment]
It should. There's already logic in place to turn a key combination into readible text, so that can be reused.

**Comment by monsdar** [comment]
Could you perhaps point out where the essential code parts are to make this feature happen? Would be a good way to get into coding for OpenRCT2.

**Comment by janisozaur** [comment]
You'd need to alter the relevant strings (at least in `en-GB.txt`, I'm not sure about other languages) to accept additional argument, then whenever a tooltip is displayed and its strings gets processed, parse hotkey information to inject into it. You can take a look at https://github.com/OpenRCT2/OpenRCT2/pull/3013.

**Comment by AaronVanGeffen** [comment]
I'd like to see this feature as well, but it may be a little troublesome to implement it, currently. All menus are implemented as regular dropdown widgets, and as such have no context for what they're representing.

In the implementation as-is, you'd have to wrap all menu items with shortcuts in a new string. This string would accepts two sub strings: the actual text, and the shortcut representing it (possibly with a `{MOVE_X}` offset, but that might be troublesome as not all menus are of the same width. You would also need to prepare the shortcut text for rendering as well.

The same is true for tooltips, although wrapping them may be more problematic w.r.t. localisation, in that the text may not flow correctly. (Not wrapping them will require localisations to be updated for sure, though, so wrapping is still preferred.)

**Comment by Broxzier** [comment]
@monsdar This piece of code converts a saved key combination to a readible string: https://github.com/OpenRCT2/OpenRCT2/blob/1148db7226c5fcf2edc4b11cb6646cf08e1daa96/src/openrct2-ui/input/KeyboardShortcuts.cpp#L124-L157

Then with `{STRING}` in the tooltips, it can be drawn from the resulting string buffer (like how janisosaur implemented https://github.com/OpenRCT2/OpenRCT2/pull/3013).

@AaronVanGeffen For the actual dropdown menus it'd be nicer to show the hotkeys on the right, like most context menus (in Windows at least). Hovering over the buttons to open a dropdown menu can have their own tooltip, right? They already have that for the rotating arrow for example ("Rotate view 90° clockwise").

**Comment by AaronVanGeffen** [comment]
> @AaronVanGeffen For the actual dropdown menus it'd be nicer to show the hotkeys on the right, like most context menus (in Windows at least).

Agreed — that's what I was getting at with the `{MOVE_X}` remark.

> Hovering over the buttons to open a dropdown menu can have their own tooltip, right? They already have that for the rotating arrow for example ("Rotate view 90° clockwise").

They can indeed. Like any tooltip, they would have to be changed accordingly. Not exactly a small undertaking...

**Comment by seifer7** [comment]
I was thinking of the exact same thing and was going to open an Issue, then found this one.
I would like to take this on if no one is currently working on it.

**Comment by AaronVanGeffen** [comment]
@seifer7 This requires a rethink of how menu items are built and drawn. Before you spend too much time on this, could you outline how you plan to implement this?

**Comment by seifer7** [comment]
@AaronVanGeffen I only had a brief look at it and can see what you mean. Looks like it'll be more involved than I first anticipated.
I'm in the middle of some other issues I'm working on, then I'll re-visit this and let you know if I come up with an idea :)

**Comment by Niels-NTG** [comment]
I've just reinstalled OpenRCT2 again after a long hiatus and I immediately was looking for this feature since this game has way too many keyboard shortcuts to easily master.

It has been 4 years when this was last discussed. Has anything changed in the UI framework that would make this feature more doable?

For instance: would it be possible to have a widget at the right-most side of dropdown menu items?

Or that isn't possible right now: would this feature still work if only tooltips had these keyboard shortcut hints? For instance: you hover over the rides toolbar button you see "Rides/attractions in park [R]", hover over research & development tab in the build rides window shows "Show research & development status [D]".

**Reply by AaronVanGeffen** [reply]
We're in the middle of reworking dropdowns. Hopefully, it will be more feasible after.

**Comment by Gymnasiast** [comment]
How should we handle shortcuts with modifiers? Those can get quite long, e.g. `CTRL + ALT + SHIFT + A`, especially with RCT’s font. I guess we could adopt macOS’ usage of `⌃⌥⇧A`, use the guest though font, or both.

**Reply by sjoerddebruin** [reply]
The macOS suggestion sounds the cleanest to me personally. How do you want to handle the command key, based on OS? I'm unaware what most Linux distro's use for it.

**Reply by AaronVanGeffen** [reply]
> The macOS suggestion sounds the cleanest to me personally.

Same, yes. It may be tricky to find common symbols, but at least the caret ^ for ctrl is fairly common, I think. 

> How do you want to handle the command key, based on OS? I'm unaware what most Linux distro's use for it.

Linux distros call it a meta or super key, and Windows calls it, well, the Windows key :-)

**Reply by sjoerddebruin** [reply]
> Linux distros call it a meta or super key, and Windows calls it, well, the Windows key :-)

I was more curious which symbol, perhaps the Windows logo as well as there are probably not much custom Linux keyboards around (mechanical for sure you can make your own decision)

**Reply by Gymnasiast** [reply]
I would use ⌘ for that, which is what macOS uses for Command, and looks enough _like_ the Windows logo without upsetting penguin aficionados everywhere ;-)

**Reply by Niels-NTG** [reply]
I don't think the game needs to have a Windows icon (or substitute) glyph anyway because as far as I know the Windows key isn't intended to be used by applications, only the OS.

**Reply by Niels-NTG** [reply]
And as for Linux I would just use `META` and `SUPER`. As far as I know there aren't really any commonly accepted symbols for these keys.

**Reply by ocalhoun6** [reply]
> I don't think the game needs to have a Windows icon (or substitute) glyph anyway because as far as I know the Windows key isn't intended to be used by applications, only the OS.

The game doesn't use it in any default shortcuts, no.

But it could be used in a player's customized shortcuts.

I doubt very many people are actually doing that, but there may be some, especially if they're using a _lot_ of custom shortcuts and were starting to run out of combinations that didn't conflict with one another.

**Comment by ColeMorda** [comment]
I've done a bit of thinking and discussion about how this could be implemented. This is a feature I would also like to see added, as I, and I'm sure many others, have trouble remembering all the useful binds for different menus.

A few questions are opened up by this idea:

#### How would you get bindings from widgets?
In order to draw the binding, you have to first know what it is. You cannot do that unless you first know what action your button does, and what binding corresponds to it. To my understanding, the `Widget` struct is used everywhere, and (in my opinion) shouldn't have any extra data added to it if only some of them are going to use that data.

One solution may be creating a map that converts `WIDX`s into their corresponding binds if applicable, but where would that go? Maybe `WindowTooltipOpen`?

Also, all current `WIDX`s would (ideally) have to be consolidated into one enum in a header file. As it stands, they are all over the place, and only in .cpp files, meaning no other files will be able to read them.

#### How should the binding tip be displayed?
It is suggested by some in the above comments that the tips should be displayed right-justified, in-line with whatever existing tooltip text exists. The problems that causes were also discussed above.

My suggestion is to add a line-break before the tip, e.g.:
```
STR_1895    :Build new ride/attraction
->
STR_1895    :Build new ride/attraction{NEWLINE}{STRING}
```
The resulting tip being:
```
-----------------------------
| Build new ride/attraction |
|     (CTRL + Alt + R)      |
-----------------------------
```

Leaving room for the (potentially very long) binding while not taking precedence over the tooltip's text itself.

#### How should the tooltip format be implemented?
The first instinct in regards to implementing the format would be to go through all the language files and append the format to every line of text that needs a binding tip. As said in https://github.com/OpenRCT2/OpenRCT2/discussions/15762#discussioncomment-1520631,
> @AaronVanGeffen:
> Not exactly a small undertaking...

And as quoted from [![Discord](https://img.shields.io/badge/discord-%23development-yellowgreen.svg)](https://discord.gg/fsEwSWs):
> @Basssiiie:
> Not sure how the devs feel about it but I would think it's best not to append those format tokens in the language files, as then all localisations also need to be updated

The same user suggests hardcoding the format in code:
> @Basssiiie:
> Wouldnt it be possible to just hardcode {NEWLINE}{STRING} in the source code and append it there if there's a shortcut?

Given the monumental task of changing the formatting of every instance of text that may need a binding hint in the language file for every single supported language, and the potential cascade of conflicts and complaints from localizers, I agree with them. Since we're already hardcoding the `WIDX`s that need binding tips, this approach wouldn't fall too far from what is already being done.

#### What about special cases?
There are quite a bit of special cases in which:
- One button may have multiple bindings that do different actions (e.g game speed up/down bindings)
- Bindings correspond to dropdown menu items which don't already contain tooltips

 As for the former, the question must be asked: "Is this binding really important enough to hint at?" If one button has multiple actions tied to it, perhaps its not worth listing all of them in a tooltip, and leaving it hidden.

As for the latter, I'm not actually sure how difficult it would be to dynamically add tooltips for these items. If they are `Widget`s, it should be as simple as checking if the tooltip's text ID is `kStringIdNone`, and just setting the tooltip string to the binding alone. Otherwise, I'm not so sure.

#### Implementation

As it stands, I think implementing something along the lines of this would be ideal (assuming all requirements mentioned above are met, and my understanding of the codebase is correct):
```c++
    static const std::unordered_map<WidgetIndex, std::string_view> sWidgetIdxToShortcutId = {
        {WIDX_CONSTRUCT_RIDE, Ui::ShortcutId::kInterfaceOpenNewRide},
...
...
    };

    static const std::string sShortcutHintFormat = "{NEWLINE}{STRING}";
    static const std::string sShortcutHintFormatEmptyTooltip = "{STRING}";

...
...
    void WindowTooltipOpen(WindowBase* widgetWindow, WidgetIndex widgetIndex, const ScreenCoordsXY& screenCoords)
    {
...
...
            result = widgetWindow->onTooltip(widgetIndex, stringId);
            // Remove early return if string doesn't exist, as the shortcut hint may still add something.
            //if (result.str == kStringIdNone)
            //    return;
        }

        // Handle shortcut hints
        auto it = sWidgetIdxToShortcutId.find(widgetIndex);
        if (it != sWidgetIdxToShortcutId.end()) {
            auto windowManager = Ui::GetWindowManager();
            std::string shortcut = windowManager->GetKeyboardShortcutString(it->second);
            /*
                Here, apply sShortcutHintFormat or its EmptyTooltip version and append it to result.
                Also requires conversion back into OpenRCT2String, somehow.
            */
        }

        if (result.str == kStringIdNone)
            return;

        WindowTooltipShow(result, screenCoords);
    }
```

**Reply by AaronVanGeffen** [reply]
Fwiw, you could just add one new string that acts as a wrapper for the existing strings, e.g. `{STRINGID}{NEWLINE}({STRING})`. That way, none of the existing strings will need to be updated, leaving less room for error.
